catalyst/loggers/wandb.py

```diff
@@ -0,0 +1,180 @@
+
+    W&B documentation: https://docs.wandb.com
+
+    Args:
+        Project: Name of the project in W&B to log to.
+        name: Name of the run in W&B to log to.
+        config: Configuration Dictionary for the experiment.
+        entity: Name of W&B entity(team) to log to.
+
+    Python API examples:
+
+    .. code-block:: python
+
+        from catalyst import dl
+
+        runner = dl.SupervisedRunner()
+        runner.train(
+            ...,
+            loggers={"wandb": dl.WandbLogger(project="wandb_test", name="expeirment_1")}
+        )
+
+    .. code-block:: python
+
+        from catalyst import dl
+
+        class CustomRunner(dl.IRunner):
+            # ...
+
+            def get_loggers(self):
+                return {
+                    "console": dl.ConsoleLogger(),
+                    "wandb": dl.WandbLogger(project="wandb_test", name="experiment_1")
+                }
+
+            # ...
+
+        runner = CustomRunner().run()
+
+    Config API example:
+
+    .. code-block:: yaml
+
+        loggers:
+            wandb:
+                _target_: WandbLogger
+                project: test_exp
+                name: test_run
+        ...
+
+    Hydra API example:
+
+    .. code-block:: yaml
+
+        loggers:
+            wandb:
+                _target_: catalyst.dl.WandbLogger
+                project: test_exp
+                name: test_run
+        ...
+    """
+
+    def __init__(
+        self, project: str, name: Optional[str] = None, entity: Optional[str] = None,
+    ) -> None:
+        self.project = project
+        self.name = name
+        self.entity = entity
+        self.run = wandb.init(
+            project=self.project, name=self.name, entity=self.entity, allow_val_change=True
+        )
+
+    def _log_metrics(self, metrics: Dict[str, float], step: int, loader_key: str, suffix=""):
+        for key, value in metrics.items():
+            self.run.log({f"{key}/{loader_key}{suffix}": value}, step=step)
+
+    def log_metrics(
+        self,
+        metrics: Dict[str, Any],
+        scope: str = None,
+        # experiment info
+        run_key: str = None,
+        global_epoch_step: int = 0,
+        global_batch_step: int = 0,
+        global_sample_step: int = 0,
+        # stage info
+        stage_key: str = None,
+        stage_epoch_len: int = 0,
+        stage_epoch_step: int = 0,
+        stage_batch_step: int = 0,
+        stage_sample_step: int = 0,
+        # loader info
+        loader_key: str = None,
+        loader_batch_len: int = 0,
+        loader_sample_len: int = 0,
+        loader_batch_step: int = 0,
+        loader_sample_step: int = 0,
+    ) -> None:
+        """Logs batch and epoch metrics to wandb."""
+        if scope == "batch":
+            metrics = {k: float(v) for k, v in metrics.items()}
+            self._log_metrics(
+                metrics=metrics, step=global_epoch_step, loader_key=loader_key, suffix="/batch"
```

Comment:
```suggestion
                metrics=metrics, step=global_batch_step, loader_key=loader_key, suffix="/batch"
```

Reply:
as far as this is per-batch/sample statistics we have to use per-sample counter

Reply:
Using `global_sample_step` will cause this problem -> https://github.com/catalyst-team/catalyst/pull/1176#issuecomment-826594360

Reply:
@AyushExel could you please try using _suffix rather than /suffix for the logging? maybe different names for the batch-based and epoch-based metrics would work for Wandb

you could not write per-**batch** metrics under the epoch counter - it's simply incorrect behaviour
could you please try using different names for per-batch and per-epoch metrics?

Reply:
Yeah, the step counter behavior is pretty strict which comes up a lot. But I'm pretty sure that's how it is right now. I tried changing names as you suggested but that didn't work. See [this ](https://docs.wandb.ai/guides/track/log#incremental-logging)where this behaviour is explained. 

> As long as you keep passing the same value for step, W&B will collect the keys and values from each call in one unified dictionary. As soon you call wandb.log() with a different value for step than the previous one, W&B will write all the collected keys and values to the history, and start collection over again. Note that this means you should only use this with consecutive values for step: 0, 1, 2, .... This feature doesn't let you write to absolutely any history step that you'd like, only the "current" one and the "next" one.

Reply:
Then I think you could use different names for batch-based and epoch-based metrics – just rename them in the way, that Wandb would understand :)

Reply:
Yes this might make more sense. So, instead of suffix, I think we can use the prefix. If you log anything in wandb as `{'prefix/key': value}`, it creates it's own section in the dashboard called `prefix` and all metrics belonging to the same prefix share the same section.

Reply:
it's up to you, you could create `{metric_name}_{prefix}/{somethin}`, so it would be accuracy_epoch and accuracy_batch

Reply:
Okay updated the names of the metrics as you suggested. Now each metric has its section like `{metric_name}_{prefix}`. I think it's all good now.